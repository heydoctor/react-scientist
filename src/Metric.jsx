// @flow

import * as React from 'react';
import PropTypes from 'prop-types';

type MetricProps = {
  on: string,
  name: string,
  data?: Object,
  options?: Object,
  style: Object,
  children: React.Node,
};

type EventArgs = {
  name: string,
  data?: Object,
  options?: Object,
};

const throwError = (message: string) => {
  throw new Error(`<Metric />: ${message}`);
};

export default class Metric extends React.Component<MetricProps> {
  node: ?HTMLDivElement;

  static defaultProps = {
    on: 'click',
    style: {},
  };

  static onEvent = (args: EventArgs) => {
    throwError(
      'This method should be overwritten somewhere before React is mounted, e.g. Metric.onEvent = () => {...}'
    );
  };

  componentDidMount() {
    if (this.node) {
      this.node.addEventListener(this.props.on, this.trackEvent);
    }
  }

  componentWillUnmount() {
    if (this.node) {
      this.node.removeEventListener(this.props.on, this.trackEvent);
    }
  }

  trackEvent = () => {
    const { name, data, options } = this.props;

    this.constructor.onEvent({ name, data, options });
  };

  render() {
    const { children, data, name, on, options, style, ...props } = this.props;

    return (
      <div
        {...props}
        className="rs-metric"
        data-testid="rs-metric"
        style={{ display: 'contents', ...style }}
        ref={ref => {
          this.node = ref;
        }}
      >
        {children}
      </div>
    );
  }
}
