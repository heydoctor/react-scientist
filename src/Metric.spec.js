import React from 'react';
import {
	render,
	fireEvent,
	cleanup,
	waitForElement,
} from 'react-testing-library';
import Metric from './Metric';

const Subject = props => (
	<Metric {...props}>
		<span>Test</span>
	</Metric>
);

afterEach(cleanup);

test('Metric.onEvent throws', () => {
	expect(() => Metric.onEvent()).toThrowErrorMatchingSnapshot();
});

describe('<Metric />', () => {
	beforeEach(() => {
		Metric.onEvent = jest.fn();
	});

	afterEach(() => {
		Metric.onEvent.mockRestore();
	});

	test('event defaults to `click`', () => {
		const name = 'Test';
		const { getByText, getByTestId, container } = render(
			<Subject name={name} />
		);

		fireEvent.click(getByTestId('rs-metric'));

		expect(Metric.onEvent).toBeCalled();
		expect(Metric.onEvent).toHaveBeenCalledWith(
			expect.objectContaining({
				name,
			})
		);
	});

	test('specify event type with `on` prop', () => {
		const name = 'Test';
		const { getByTestId } = render(<Subject name={name} on="mouseover" />);

		fireEvent.mouseOver(getByTestId('rs-metric'));

		expect(Metric.onEvent).toBeCalled();
		expect(Metric.onEvent).toHaveBeenCalledWith(
			expect.objectContaining({
				name,
			})
		);
	});

	test('sends along data', () => {
		const name = 'Test';
		const data = {
			hello: 'world',
		};
		const { getByTestId } = render(<Subject name={name} data={data} />);

		fireEvent.click(getByTestId('rs-metric'));

		expect(Metric.onEvent).toHaveBeenCalledWith(
			expect.objectContaining({
				name,
				data,
			})
		);
	});

	test('renders children', () => {
		const { getByText } = render(<Subject name="Test" />);
		const child = getByText('Test');

		expect(child.tagName).toEqual('SPAN');
		expect(child.textContent).toEqual('Test');
	});
});
