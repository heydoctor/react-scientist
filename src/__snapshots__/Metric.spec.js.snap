// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Metric.onEvent throws 1`] = `"<Metric />: This method should be overwritten somewhere before React is mounted, e.g. Metric.onEvent = () => {...}"`;
